DM-13365: Proper Caching for remote object stores

timj opened the pull request:
Includes cache expiry.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add internal registry of cached files with cache expiration

Supports: files, datasets, age, and size expiration.

timj committed:
Include StorageClass of composite in list when given component

The composite dataset type name was already being included and
it was confusing for datastore caching to depend on whether
composite disassembly is enabled.

We may want to revisit this by having an explicit "include_parent"
parameter in lookupNames method.

timj committed:
Remove temporary cache directory at process exit

timj committed:
Ensure that cache gets component-ized dataset ref

Previously it did not matter because the URI being read already
contained the component information. The cache though is creating
a new filename and has to know that a component is involved so
that the names can be unique.

timj committed:
Add ButlerURI.temporary_uri

This can be used to create local or remote temporary URIs.

timj committed:
Use ButlerURI.temporary_uri for managing remote store temporary

This allows the cache to move the file somewhere else without
complaint and is more consistent with ButlerURI usage elsewhere.

timj committed:
Include component name in cache file name

The dataset ref ID is common to all components when a composite
is disassembled so the component is required to ensure that the
filename is unique.

timj committed:
Short circuit cache search if we know there is no cache directory

This also prevents the directory from being created too early.

timj committed:
Add remove_from_cache method and trigger it on datastore.trash

For some tests in particular we prune a dataset and then
immediately put a new one. If that dataset was cached and
not removed from the cache the cache infrastructure would
raise.

timj committed:
Add ability to control default caching status

timj committed:
Allow cache move to overwrite existing file

This can be possible if two butlers are retrieving the same
dataset into a shared cache directory.

timj committed:
Pass a ref to the cache to take into account composites

In some cases a ref alone is not enough information for the
cache to work out whether it is caching a standalone component
or a composite. The datastore does know, so pass that information
to the cache.

timj committed:
Improve description of cache expiry

timj committed:
Allow environment variables to override the cache config settings

timj committed:
Add __str__ for cache to report status

timj committed:
Use cache even when reading bytes from datastore

If a dataset has been told to cache then it should be more
efficient to read the bytes directly from local cache than downloading
them from object store.

timj committed:
Fix disassembly test template to include component

Otherwise some tests fail because the files are not unique
when disassembled.

timj committed:
Change find_in_cache method to be context manager

This allows the cache to protect the file against being deleted
by another process using the shared cache and clearing the file.

timj committed:
Add file_count and cache_size attributes to cache object

timj committed:
Add tests for cache expiry

timj committed:
Add test for composite disassembly and datasets expiration

timj committed:
Set some reasonable defaults for cache expiry

Just cache up to 4 Exposure-derived datasets.
No longer need to cache the stray light data.

timj committed:
Cache downloaded file after we've finished with it

This removes any possibility that another process will remove
it from the cache whilst this process is reading it.

timj commented on python/lsst/daf/butler/configs/datastores/fileDatastore.yaml:
```diff
@@ -19,13 +19,23 @@ datastore:
       # The formatter associated with this storage class
       # implements deferred loading of data
       StrayLightData: true
+      ExposureF: true
```
Once StrayLightData is turned into a normal dataset we need to decide whether it's still worth caching. We may want to add `Exposure` to this list in addition to `ExposureF` (or fix things so that storage class inheritance works in lookups). Not sure what other large non-exposure datasets we'd like to cache.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -724,6 +732,54 @@ def as_local(self) -> Iterator[ButlerURI]:
+        ----------
+        prefix : `ButlerURI`, optional
+            Prefix to use. Without this the path will be formed as a local
+            file URI in a temporary directory.
```
This makes it sound like it might be in a `mkdtemp` directory (and perhaps have user umask) rather than being in an OS-specified one.  Also note that if the prefix contains a directory, that must exist (and be writable).

ktlim commented on python/lsst/daf/butler/core/_butlerUri/_butlerUri.py:
```diff
@@ -724,6 +732,54 @@ def as_local(self) -> Iterator[ButlerURI]:
+            # To avoid inconsistencies in behavior, delete the file
+            # even though that might result in a name clash later.
+            try:
+                temporary_uri.remove()
```
This seems less than wonderful.  If you're not going to use any of the features of `mkstemp`, maybe it would be better to just create a random filename in `tempfile.gettempdir()`?

timj replied:
You mean because I am already doing that logic in the `else` block below for the remote URI?

ktlim replied:
Yes, if it works for one, it should work for the other.

timj replied:
I've rewritten this to not involve mkstemp.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1070,11 +1067,20 @@ def _removeFileExists(uri: ButlerURI) -> None:
                 uri.write(serializedDataset, overwrite=True)
                 log.debug("Successfully wrote bytes directly to %s", uri)
 
+                # It may be that this dataset should be cached locally
```
Is this actually more efficient than writing to temp and then copying?  In particular, the serialization code might be inefficient, so executing it twice would be bad.

It's also a little confusing having the main code for `toBytes` being here in an `else:`, though I understand catching `NotImplementedError` is Pythonic.

I guess this could be rewritten as:
```python
if not self.cacheManager.should_be_cached(ref):
    try:
        toBytes()
    except NotImplementedException:
        pass
    else:
        uri.write()
        return self._extractIngestInfo()
# rest of write-to-temp-and-copy code goes here
```

 (Would be even nicer to be able to stream from the serializer to the URI, but that's another refactor or two down the road.)

timj replied:
It's not serializing the in-memory dataset twice -- the bytes are just bytes that are written twice. Rewriting it so that we always write to the local file first if the intent is to be cached seems reasonable (I hadn't really intended to cache these at all initially -- the tests all failed without because the tests all use very small datasets that can bypass the file writing stage -- hence the convoluted logic I ended up with). 

There is a related problem that if the temp directory is on a different file system to the cache directory we will have to (effectively) read all the bytes back into memory for the "move" since `os.rename` won't work.

timj replied:
Thanks. The rewrite is much clearer.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1085,6 +1091,14 @@ def _read_artifact_into_memory(self, getInfo: DatastoreFileGetInformation,
+            This ref must have the same ID as the supplied ref but can
+            be a parent ref or component ref to indicate to the cache whether
+            a composite file is being requested from the cache or a component
+            file. Without this the cache will default to the suppled ref but
```
```suggestion
            file. Without this the cache will default to the supplied ref but
```

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1137,45 +1164,53 @@ def _read_artifact_into_memory(self, getInfo: DatastoreFileGetInformation,
+                            # it until after we've used it (to ensure it can't
+                            # be expired whilst we are using it).
+                            can_be_cached = True
+                            cache_msg = " and likely cached"
```
Why "likely" here?  Isn't it guaranteed, at least initially?  So "will be"?

timj replied:
I think I say "likely" because if the `formatter.read` fails I will not be caching it.

ktlim commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1137,45 +1164,53 @@ def _read_artifact_into_memory(self, getInfo: DatastoreFileGetInformation,
-                                     f" ({ref.datasetType.name} from {uri}): {e}") from e
+                with uri.as_local() as local_uri:
+
+                    # URI was remote and file was downloaded
```
I think this should move next to the `if`, perhaps even after it.

ktlim commented on tests/test_datastore.py:
```diff
@@ -1056,7 +1061,26 @@ def setUp(self):
 
         # Create empty files
```
Not empty anymore, are they?

ktlim commented on tests/test_datastore.py:
```diff
@@ -1056,7 +1061,26 @@ def setUp(self):
-            uri.write(b"")
+            uri.write(b"0123456789")
+
+        # Create a some composite refs with component files.
```
```suggestion
        # Create some composite refs with component files.
```

ktlim commented on tests/test_datastore.py:
```diff
@@ -1115,20 +1152,155 @@ def assertCache(self, cache_manager):
+        with cache_manager.find_in_cache(self.refs[2], ".txt") as found:
+            self.assertIsNotNone(found)
+
+        # But not the dataset that was stored after it.
```
Isn't this comment backwards?  You keep both the file and the one after, since it's not `None`.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -31,27 +31,194 @@
+        - "component": The name of the component (can be `None`),
+        - "extension": File extension (can be `None`).
+    """
+    root_ext = cached_location.split(".", maxsplit=1)
```
Is `rsplit()` safer?  Or even `root, _, ext = cached_location.rpartition(".")` ?

Or even `re.search("(.+?)(_([^_]+?))?(\..+)")` ?

timj replied:
You are worried about rogue files being found in the cache directory (especially for a shared cache)?

ktlim replied:
I was worried about the `id` field, but since you control that, there's no problem.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -31,27 +31,194 @@
+    ref : `DatasetRef`
+        The dataset to look up in or write to the cache.
+    extension : `str`
+        File extension to use for this file.
```
I guess you're consistent everywhere, but maybe note that "." must be included in the extension?

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -31,27 +31,194 @@
+    # disassembly is happening since the ID is shared for all components.
+    component = ref.datasetType.component()
+    component = f"_{component}" if component else ""
+    return root.join(f"{ref.id}{component}{extension}")
```
Are you sure `ref.id` can't contain `_`?

timj replied:
It can only be a UUID or integer so yes, none of those have underscores. Would you like an `assert` to be included?

ktlim replied:
Oh, if this is from a field you control, then no problem.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -31,27 +31,194 @@
+    return {"id": id_, "component": component, "extension": ext}
+
+
+class CacheEntry(BaseModel):
```
Are you basically using `BaseModel` everywhere now instead of `dataclass`?

timj replied:
Yes, I am generally doing that and in this case it's only creating one of them so there's no performance hit.

timj replied:
Sorry -- I was thinking CacheRegistry not CacheEntry -- yes, there will be a bit of overhead for the validation but it seemed reasonable and the validation is very simple.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -153,32 +355,104 @@ class DatastoreCacheManager(AbstractDatastoreCacheManager):
+    * ``$DAF_BUTLER_CACHE_DIRECTORY``
+    * ``$DAF_BUTLER_CACHE_EXPIRATION_MODE``
+
+    The expiration mode should take the form a ``mode=threshold`` so for
```
```suggestion
    The expiration mode should take the form ``mode=threshold`` so for
```

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -153,32 +355,104 @@ class DatastoreCacheManager(AbstractDatastoreCacheManager):
+
+        if self._cache_directory:
+            if not self._cache_directory.isLocal:
+                raise ValueError(f"Cache directory must be a file system. Got: {self._cache_directory}")
```
```suggestion
                raise ValueError(f"Cache directory must be on a local file system. Got: {self._cache_directory}")
```

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -31,27 +31,194 @@
+    name: str
+    """Name of the file."""
+
+    size: int
```
I'm wondering if it isn't almost as easy to get these from the filesystem each time rather than maintaining a possibly-out-of-sync in-memory copy.

But I guess it's easier to figure out datasets by id this way.

timj replied:
The file sizes can't change so they are static information. Also I was under the impression that file systems don't really like repeated calls to `stat` if the cache grows big so limiting the cache check to a `readdir` rather than a `readdir`+`stat` seemed like a reasonable thing to do.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -223,21 +497,277 @@ def move_to_cache(self, uri: ButlerURI, ref: DatasetRef) -> Optional[ButlerURI]:
+
+            # If the file happens to also exist in the exempt folder
+            # skip removal since a process is actively using the file.
+            if self._temp_exempt_directory.join(entry).exists():
```
I'm worried that there still may be (brief but non-zero) windows where a race can happen.  Still, since you treat it as uncached in that case, it should be safe if less performant.

timj replied:
I did think about using hard links exclusively and letting the cache expiration delete the file regardless. Then I worried that someone might put the cache dir on some file system where hard links wouldn't work for some reason and so I played it a bit safer by skipping the removal if it existed in both places. I'm not sure how to deal with the case of another process deleting the file in the instant before I create a soft link in the exempt directory (the hardlink race would cause the hardlink to fail to be made but would helpfully fall back to creating a soft link to a file that is no longer there).

timj replied:
Do you want to block merging until we try to make this robust? How much can we rely on hardlinks working within the temp directory? Hardlinks working mean that we can remove this specific test completely and try to make a hardlink. If the hardlink is made then there is no race (deleting the cache file won't delete the thing in the "exempt" directory). If the hardlink fails we will have to report that the file wasn't really cached. That is all great assuming hardlinks can be trusted. Softlinks are the reason there is a possible race.

timj replied:
hardlinks are much cleaner because then the cache directory can expire as normal without having to keep the extra files around and the "exempt" directory can delete things independently (let's hope they aren't kept around for too long if disk space is tight)

ktlim replied:
Since these are on the worker and all within the same directory tree, I think hardlinks are reliable.  But having any race be "uncached" is also a belt-and-suspenders safety precaution.

timj replied:
I've removed the softlink option which also removes any need for cache expiry to know about the "exempt" directory. The cache doesn't have to be on the worker -- people can specify a directory that is on some shared file system if they want.

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -223,21 +497,277 @@ def move_to_cache(self, uri: ButlerURI, ref: DatasetRef) -> Optional[ButlerURI]:
+            return
+
+        if self._expiration_mode == "datasets":
+            # Count the datasets, using reverse date order
```
Isn't this ascending timestamp order, not "reverse date order"?

ktlim commented on python/lsst/daf/butler/core/datastoreCacheManager.py:
```diff
@@ -223,21 +497,277 @@ def move_to_cache(self, uri: ButlerURI, ref: DatasetRef) -> Optional[ButlerURI]:
+            if n_over > 0:
+                keys_to_remove = []
+                removed = 0
+                for dataset in datasets:
```
Can't you just iterate over `datasets[:n_over]` like above, skipping the whole `removed` thing?

timj replied:
Simplified.

timj committed:
Add news fragment

timj committed:
Fix typo in docstring

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Make error message clearer

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Typo fix in python/lsst/daf/butler/datastores/fileDatastore.py

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Update comment in tests/test_datastore.py

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

timj committed:
Rewrite temporary_uri to unify code paths

Use mkstemp for local and self-generated for remote was an added
complication without much benefit given that we immediately delete
the temp file and so lose the protection for overwriting that
is provided by mkstemp.

timj committed:
Be explicit about extension including the leading "."

timj committed:
Fix some comments

timj committed:
Fix the caching test

Do an extra caching request to ensure that the file would have
been removed but wasn't.

timj committed:
Rearrange remote datastore file writing to simplify logic

* Always use tempfile route if the intent is to cache the dataset,
  even if direct serialization to bytes is possible.
* Unpack the try/except hierarchy such that a failure to in the
  formatter writing a file does not show an error stack that
  includes the (expected) failure to serialize to bytes in memory.

Also change the test configuration such that webdav and s3
cache some datasets but not others and so can test the code
path for both direct bytes and file writing.

timj committed:
Simplify code logic for dataset expiration

timj commented:
@ktlim I've made all the requested changes. The one outstanding issue is worrying about the exemption race and whether we can force hardlinks and stop trying to fall back to symlinks.

timj committed:
Remove softlink option in cache retrieval

Hard links protect against races of other processes doing cache
expiry just after the soft link has been created. A hard link
solves this because either the hard link is created or it isn't
created.

If it is created the file in the exempt directory can be used
without risk of being deleted prematurely and it is safe to
delete the version in the main cache. If the hard link can't
be created because another process has just deleted the
file, it will be as if the file wasn't in the cache at all.

Assumes that hard links will work on any file system we use
for the cache.

timj committed:
Python 3.8 or higher

timj committed:
Mention ADASS paper

timj committed:
Add codeql action

This is not a required action but is used to experiment.
